sarif/src/main/java/com/jetbrains/qodana/sarif/RuleUtil.java

```diff
@@ -0,0 +1,47 @@
+package com.jetbrains.qodana.sarif;
+
+import com.jetbrains.qodana.sarif.model.*;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.Collection;
+import java.util.Objects;
+import java.util.stream.Stream;
+
+public class RuleUtil {
+    private RuleUtil() {
+        throw new IllegalStateException("No instances.");
+    }
+
+    @Nullable
+    public static ReportingDescriptor findRuleDescriptor(@NotNull SarifReport report, @Nullable String ruleId) {
+        if (ruleId == null || ruleId.isEmpty()) return null;
+
+        return allRules(report)
+                .filter(r -> Objects.equals(r.getId(), ruleId))
+                .findFirst()
+                .orElse(null);
+    }
+
+    @NotNull
+    public static Stream<ReportingDescriptor> allRules(@NotNull SarifReport report) {
+        return stream(report.getRuns())
```

Comment:
If RuleUtil#findRuleDescriptor will be called multiple times, this pretty expensive method will be also called multiple times. WDYT about transofrming RuleUtil in a class with a state. Indeed, this will be used for an unchanged report, thus making it possible to instantiate it for such a report and persisting allRules once computed.

Reply:
Yep, it is indeed very expensive - and I somehow thought that the resolution will take place only after the ruleIds have been made distinct. Good catch.

My train of thought regarding state/no state was that the `DescriptorLookup` is not a public class of this library, so it can kinda "ignore" the general problem of mutability in our report classes. 
However, `RuleUtil` _is_ public, so I would not expose state which is not guaranteed to be correct. I'd rather implement it here on use-site.

Reply:
Updated to memoziation on use-site